Enhanced Model Management and SystemPrompt Integration in OllamaAgent

cnupy opened the pull request:
# PR Summary

## Feature 1: Support for Multiple Models 
### Description
OllamaAgent can now handle and switch between multiple models seamlessly, listing available models dynamically via an API
### Implementation Details
*   Modified the agent to include a model selection mechanism that allows it to iterate through predefined models.
*   Introduced a `model` command to manage the switching between models based on available models.

## Feature 2: Integration of SystemPrompt 
### Description
It is possible to configure SystemPrompt with the selected model into OllamaAgent, providing users with a toolkit for task-oriented conversations.
### Implementation Details
*   Modified the agent to accommodate Ollama's SystemPrompt functionality.
*   Introduced a `system-prompt` command to manage the setting.

## Feature 3: Support for Predefined ModelConfigs sets 
### Description
OllamaAgent now supports the use of predefined model configurations, allowing users to easily switch between different models and system prompts based on specific requirements.
### Implementation Details
*   Introduced a `ModelConfig` Record to encapsulate data for each configuration.
*   Introduced a `config` command to manage the switching between configuration based on available predefined sets.

<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

<!-- Summarize your PR between here and the checklist. -->

## PR Context

<!-- Provide a little reasoning as to why this Pull Request helps and why you have opened it. -->
This Pull Request enhances the functionality of OllamaAgent by introducing improved model management. By allowing seamless switching between multiple models and supporting predefined configurations, it offers users greater flexibility. The addition of SystemPrompt functionality empowers users with a more task-oriented conversational toolkit, streamlining interactions. These updates not only enhance user experience but also improve the agent's adaptability to diverse requirements. This makes OllamaAgent a more versatile and powerful tool overall.

cnupy commented:
> @cnupy please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement
> ### Contribution License Agreement
> This Contribution License Agreement (**“Agreement”**) is agreed to by the party signing below (**“You”**), and conveys certain license rights to Microsoft Corporation and its affiliates (“Microsoft”) for Your contributions to Microsoft open source projects. This Agreement is effective as of the latest signature date below.
> 
> 1. **Definitions**.
>    **“Code”** means the computer software code, whether in human-readable or machine-executable form,
>    that is delivered by You to Microsoft under this Agreement.
>    **“Project”** means any of the projects owned or managed by Microsoft and offered under a license
>    approved by the Open Source Initiative ([www.opensource.org](http://www.opensource.org)).
>    **“Submit”** is the act of uploading, submitting, transmitting, or distributing code or other content to any
>    Project, including but not limited to communication on electronic mailing lists, source code control
>    systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
>    discussing and improving that Project, but excluding communication that is conspicuously marked or
>    otherwise designated in writing by You as “Not a Submission.”
>    **“Submission”** means the Code and any other copyrightable material Submitted by You, including any
>    associated comments and documentation.
> 2. **Your Submission**. You must agree to the terms of this Agreement before making a Submission to any
>    Project. This Agreement covers any and all Submissions that You, now or in the future (except as
>    described in Section 4 below), Submit to any Project.
> 3. **Originality of Work**. You represent that each of Your Submissions is entirely Your original work.
>    Should You wish to Submit materials that are not Your original work, You may Submit them separately
>    to the Project if You (a) retain all copyright and license information that was in the materials as You
>    received them, (b) in the description accompanying Your Submission, include the phrase “Submission
>    containing materials of a third party:” followed by the names of the third party and any licenses or other
>    restrictions of which You are aware, and (c) follow any other instructions in the Project’s written
>    guidelines concerning Submissions.
> 4. **Your Employer**. References to “employer” in this Agreement include Your employer or anyone else
>    for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
>    Submission is made in the course of Your work for an employer or Your employer has intellectual
>    property rights in Your Submission by contract or applicable law, You must secure permission from Your
>    employer to make the Submission before signing this Agreement. In that case, the term “You” in this
>    Agreement will refer to You and the employer collectively. If You change employers in the future and
>    desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
>    and secure permission from the new employer before Submitting those Submissions.
> 5. **Licenses**.
> 
> * **Copyright License**. You grant Microsoft, and those who receive the Submission directly or
>   indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license in the
>   Submission to reproduce, prepare derivative works of, publicly display, publicly perform, and distribute
>   the Submission and such derivative works, and to sublicense any or all of the foregoing rights to third
>   parties.
> * **Patent License**. You grant Microsoft, and those who receive the Submission directly or
>   indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license under
>   Your patent claims that are necessarily infringed by the Submission or the combination of the
>   Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
>   import or otherwise dispose of the Submission alone or with the Project.
> * **Other Rights Reserved**. Each party reserves all rights not expressly granted in this Agreement.
>   No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
>   granted by implication, exhaustion, estoppel or otherwise.
> 
> 6. **Representations and Warranties**. You represent that You are legally entitled to grant the above
>    licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
>    have disclosed under Section 3). You represent that You have secured permission from Your employer to
>    make the Submission in cases where Your Submission is made in the course of Your work for Your
>    employer or Your employer has intellectual property rights in Your Submission by contract or applicable
>    law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
>    have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
>    You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
>    REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
>    EXPRESSLY STATED IN SECTIONS 3, 4, AND 6, THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
>    PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
>    NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
> 7. **Notice to Microsoft**. You agree to notify Microsoft in writing of any facts or circumstances of which
>    You later become aware that would make Your representations in this Agreement inaccurate in any
>    respect.
> 8. **Information about Submissions**. You agree that contributions to Projects and information about
>    contributions may be maintained indefinitely and disclosed publicly, including Your name and other
>    information that You submit with Your Submission.
> 9. **Governing Law/Jurisdiction**. This Agreement is governed by the laws of the State of Washington, and
>    the parties consent to exclusive jurisdiction and venue in the federal courts sitting in King County,
>    Washington, unless no federal subject matter jurisdiction exists, in which case the parties consent to
>    exclusive jurisdiction and venue in the Superior Court of King County, Washington. The parties waive all
>    defenses of lack of personal jurisdiction and forum non-conveniens.
> 10. **Entire Agreement/Assignment**. This Agreement is the entire agreement between the parties, and
>     supersedes any and all prior agreements, understandings or communications, written or oral, between
>     the parties relating to the subject matter hereof. This Agreement may be assigned by Microsoft.

@microsoft-github-policy-service agree

kborowinski commented on shell/agents/AIShell.Ollama.Agent/OllamaAgent.cs:
```diff
@@ -238,15 +250,23 @@ public async Task<bool> ChatAsync(string input, IShell shell)
         {
             host.WriteErrorLine($"{e.Message}");
-            host.WriteErrorLine($"Ollama model:    \"{_settings.Model}\"");
+            host.WriteErrorLine($"Ollama active model:    \"{activeModel.Name}\"");
```
@cnupy 
This is just cosmetics, but I would change the indentation as follows:
```powershell
host.WriteErrorLine($"Ollama active model: \"{activeModel.Name}\"");
host.WriteErrorLine($"Ollama endpoint:     \"{_settings.Endpoint}\"");
host.WriteErrorLine($"Ollama settings:     \"{SettingFile}\"");
```

cnupy replied:
Thanks! Fixed!

daxian-dbw commented:
@kborowinski @cnupy After upgrading to the v0.5.13 of Ollama, the `OllamaAgent` stops working even from the main branch ... After sending a query, it's just spinning forever, and after cancel the request, the server side displays the following record:

![image](https://github.com/user-attachments/assets/86f84433-7298-4f97-aaa7-971e5b430531)

I moved to `OllamaSharp` [v5.1.4](https://www.nuget.org/packages/OllamaSharp/5.1.4), but it's the same result. Any idea what could be the problem?

kborowinski commented:
@daxian-dbw I'm on latest `AIShell` build from *main* and `Ollama v0.5.13` with `llama3.1:8b-instruct-q5_K_M` and it works. What model are you running?

![Animation](https://github.com/user-attachments/assets/ce03cab8-7a8a-494d-91fc-f158f5b55184)

cnupy commented:
Everything's working fine for me as well. I'm using Ollama 0.5.4
```shell
/llm/ollama# ./ollama -v
ggml_sycl_init: found 1 SYCL devices:
ollama version is 0.5.4-ipexllm-20250310

cnupy commented on shell/agents/AIShell.Ollama.Agent/Settings.cs:
```diff
@@ -1,38 +1,219 @@
+
+                if (string.IsNullOrEmpty(this.RunningConfig.ModelName))
+                {
+                    // Active GPT not specified, but there is only one GPT defined, then use it by default.
```
Fixed.

cnupy commented on shell/agents/AIShell.Ollama.Agent/Settings.cs:
```diff
@@ -1,38 +1,219 @@
+
+        RunningConfig = new RunningConfig();
+
+        if (Configs is not null)
```
Thanks for the suggestion! I've refactored the code, and it's much cleaner now.

cnupy commented on shell/agents/AIShell.Ollama.Agent/Settings.cs:
```diff
@@ -1,38 +1,219 @@
+    private async Task EnsureModelsInitialized(CancellationToken cancellationToken = default)
+    {
+        if (!_initialized)
+        {
```
Thanks!

cnupy commented on shell/agents/AIShell.Ollama.Agent/Settings.cs:
```diff
@@ -1,38 +1,219 @@
+        host.RenderList(
+            RunningConfig.SystemPrompt,
+            [
+                new CustomElement<string>(label: "System prompt is", str => str)
```
Fixed. The semicolon is already included when using CustomElement render element.

cnupy commented on shell/agents/AIShell.Ollama.Agent/Settings.cs:
```diff
@@ -1,38 +1,219 @@
+    {
+        await EnsureModelsInitialized(cancellationToken).ConfigureAwait(false);
+        host.RenderTable(
+            [.. _models],
```
No particular reason, but nice suggestion. Fixed.

cnupy commented on shell/agents/AIShell.Ollama.Agent/Settings.cs:
```diff
@@ -1,38 +1,219 @@
+                if (string.IsNullOrEmpty(this.RunningConfig.ModelName))
+                {
+                    // Active GPT not specified, but there is only one GPT defined, then use it by default.
+                    _activeModel = _models.First();
```
This case is now handled a bit differently after I refactored the Settings class. Thanks again for the suggestion!

daxian-dbw commented on shell/agents/AIShell.Ollama.Agent/OllamaAgent.cs:
```diff
@@ -238,15 +250,23 @@ public async Task<bool> ChatAsync(string input, IShell shell)
+        }
+        finally
+        {
+            if (_settings.RunningConfig.ResetContext)
+            {
+                // Reset the request context
+            	ResetContext();
+            }
```
The XML comment for `ResetContext` says:
> Indicates whether the context should be reset when switching to this configuration.

It sounds to me we should reset context only when switching to a new config. But here you are resetting the context for every chat.

cnupy replied:
The comment is incorrect. The intended behavior is to reset the context with each interaction. However, I am still unsure about this option. My intention was for it to reset the context automatically, but it appears that Ollama does not honor this setting as expected, retaining the context even when it's set to null. I may need to consider removing this option.

There is a better way to handle this. I am currently experimenting with using the OllamaApiClient  method instead of . However, this will be addressed in a future pull request.

daxian-dbw replied:
In that case, maybe remove the `ResetContext` setting in this PR?

cnupy replied:
Yes, removed.

daxian-dbw commented on shell/agents/AIShell.Ollama.Agent/Command.cs:
```diff
@@ -0,0 +1,278 @@
+internal sealed class ConfigCommand : CommandBase
+{
+    private readonly OllamaAgent _agnet;
+     public ConfigCommand(OllamaAgent agent)
```
```suggestion

    public ConfigCommand(OllamaAgent agent)
```

daxian-dbw reviewed (commented):
Sorry for the delay!
The ollama agent from the main branch works for me now (I have no idea why it didn't weeks before. I guess something specific to my system). I left some comments, but haven't finished reviewing the `Command.cs` file. Will finish it tomorrow.

daxian-dbw commented on shell/agents/AIShell.Ollama.Agent/Command.cs:
```diff
@@ -0,0 +1,278 @@
+    }
+}
+
+internal sealed class ModelCommand : CommandBase
```
nitpick: we retrieve the available models from Ollama only once (even if there is changes to the setting file), so in case a new model is pulled on the Ollama side, a user has to restart the Ollama agent to be able to see it. How about make the `model` command always refresh `_settings._availableModels`? So that the user would see the new model when using the `model` command.

cnupy replied:
I am currently working on the second iteration of the implementation and wanted this version merged before pushing additional changes. In this iteration, I am adding commands to pull, create, and delete models directly from AIShell. These commands will utilize a live model list and leverage the `ChatAsync` method instead of `GenerateAsync`.

daxian-dbw replied:
Sounds good. Then let's ignore this comment for now :)

daxian-dbw reviewed (commented):
Done with my 2nd pass of review :)

cnupy commented on shell/agents/AIShell.Ollama.Agent/Settings.cs:
```diff
@@ -1,40 +1,193 @@
 {
-    public string Model { get; }
+    private bool _initialized = false;
+    private List<string> AvailableModels { get; set; } = [];
```
Good catch. Accepted.

cnupy commented on shell/agents/AIShell.Ollama.Agent/Settings.cs:
```diff
@@ -1,40 +1,193 @@
+            return;
+        }
+
+        OllamaApiClient _client = null;
```
Accepted.

cnupy commented on shell/agents/AIShell.Ollama.Agent/Settings.cs:
```diff
@@ -1,40 +1,193 @@
+        {
+            _client = new OllamaApiClient(Endpoint);
+            var models = await _client.ListLocalModelsAsync(cancellationToken).ConfigureAwait(false);
+            AvailableModels = [.. models.Select(m => m.Name)];
```
They are listed with the tags and the tag must be specified with the `/mode use` command:
```@ollama> /model list

  Model Name                          Active
──────────────────────────────────────────────
  qwen2.5-powershell-expert:3b
  qwen2.5:3b
  llama3.2-powershell-expert:3b
  llama3.2:3b
  llama3.2:latest

@ollama> /model use llama3.2
ERROR: A model with the name 'llama3.2' doesn't exist in the list of available models. Available Modless: qwen2.5-powershell-expert:3b, qwen2.5:3b,        
llama3.2-powershell-expert:3b, llama3.2:3b, llama3.2:latest.
  ```

daxian-dbw replied:
Today, a user can specify `"Model": "phi3"` in the `ollama.config.json` file, and it works fine when assigning `phi3` directly to `_request.Model`. I guess `OllamaSharp` tries expanding `phi3` to `phi3:latest`, like when you use a docker container image. With the change in this PR, `"Model": "phi3"` stops working -- default setting file created by `NewExampleSettingFile()` throws the following exception for me:

```
ERROR: Agent failed to generate a response: The configured Ollama model 'phi3' doesn't exist in the list of available models.
   at AIShell.Ollama.Agent.Settings.GetActiveModel(IHost host, CancellationToken cancellationToken) in
E:\arena\repos\PSCopilot\shell\agents\AIShell.Ollama.Agent\Settings.cs:line 154
```
Would it make sense to keep the current behavior? If the module name contains `:`, then simply compare it; if not, then append `:latest` and compare.

cnupy replied:
Ah, of course. I didn’t understand your question right away.

cnupy commented on shell/agents/AIShell.Ollama.Agent/Settings.cs:
```diff
@@ -1,40 +1,193 @@
+        }
+        finally
+        {
+            if (_client is IDisposable disposable)
+            {
+                disposable.Dispose();
+            }
```
Thanks for the idea!

cnupy commented on shell/agents/AIShell.Ollama.Agent/Settings.cs:
```diff
@@ -1,40 +1,193 @@
+        await EnsureModelsInitialized(cancellationToken).ConfigureAwait(false);
+        if (!AvailableModels.Contains(name))
+        {
+            throw new InvalidOperationException($"A model with the name '{name}' doesn't exist in the list of available models.");
```
Thanks.

cnupy commented on shell/agents/AIShell.Ollama.Agent/Settings.cs:
```diff
@@ -1,40 +1,193 @@
+    internal async Task UseConfg(ModelConfig config, CancellationToken cancellationToken = default)
+    {
+        RunningConfig = config with { };
+        await UseModel(RunningConfig.ModelName, cancellationToken).ConfigureAwait(false);
```
I did some refactoring. It should be better now.

daxian-dbw commented on shell/agents/AIShell.Ollama.Agent/Settings.cs:
```diff
@@ -1,40 +1,193 @@
+        }
+        else
+        {
+            if (!AvailableModels.Contains(RunningConfig.ModelName))
```
That makes sense. However, in `ChatAsync`, we will be calling `_settings.GetActiveModel(host)` per every query, and we will be checking `!_availableModels.Contains(RunningConfig.ModelName)` every time. That seems to be wasted efforts that can be avoided.

cnupy replied:
Got it now! I've added a flag to ensure the check is performed only once.

daxian-dbw commented:
@cnupy Thanks for the quick updates. Now only those 4 remaining comments are relevant. All rest looks good.

cnupy commented on shell/agents/AIShell.Ollama.Agent/Settings.cs:
```diff
@@ -1,40 +1,193 @@
+/// <summary>
+/// Represents a configuration for an Ollama model.
+/// </summary>
+/// <param name="Name">Required. The unique identifier name for this configuration.</param>
+/// <param name="ModelName">Required. The name of the Ollama model to be used.</param>
```
I've added the `[JsonRequired]` attribute to ensure the required keys are properly marked.

cnupy commented:
> @cnupy Thanks for the quick updates. Now only those 4 remaining comments are relevant (I left follow-up comments). All rest looks good.

I've resolved the four remaining comments. Additionally, I renamed the `Configs` property to `Presets` for better clarity. I've also carried out some additional refactoring to properly implement the running model checking. Furthermore, I moved and optimized `PerformSelfcheck` from `OllamaAgent` to `Settings` so that it can be utilized in both locations. Let me know if there's anything else you'd like adjusted.

daxian-dbw commented:
@cnupy To be more efficient, I went ahead and made some refactoring directly. The main changs are:
1. Update `PerformSelfcheck` to account for `RunningConfig` check as well, while making it possible to do endpoint check only too. With this change, we will only need a call to `PerformSelfcheck`, and then we can directly use `_settings.RunningConfig.ModelName`.
2. Update `EnsureModelsInitialized` to not do the endpoint check when the passed-in host is null.
3. Use `async` actions for commands where they are applicable.

Please review my changes and see if you have any concerns. Thanks!

daxian-dbw commented:
BTW, the Ollama server writes out a warning about using the `context` field:
> [GIN] 2025/04/23 - 13:05:47 | 200 |    9.6578815s |       127.0.0.1 | POST     "/api/generate"
time=2025-04-23T13:06:59.988-07:00 level=WARN source=routes.go:274 msg="the context field is deprecated and will be removed in a future version of Ollama"

So, the ollama agent code will need to be updated accordingly to avoid breaking.

cnupy commented on shell/agents/AIShell.Ollama.Agent/Settings.cs:
```diff
@@ -1,40 +1,268 @@
+
+        if (!checkEndpointOnly && !_runningConfigChecked)
+        {
+            await EnsureModelsInitialized(host: null).ConfigureAwait(false);
```
```suggestion
            await EnsureModelsInitialized(host: host, performSelfcheck: false).ConfigureAwait(false);
```

cnupy commented on shell/agents/AIShell.Ollama.Agent/Command.cs:
```diff
@@ -0,0 +1,295 @@
+                    CancellationToken.None);
+            }
+
+            await settings.UseModel(host: null, name);
```
```suggestion
            await settings.UseModel(host: host, name: name);
```

cnupy commented on shell/agents/AIShell.Ollama.Agent/Settings.cs:
```diff
@@ -1,40 +1,268 @@
+
+    internal async Task UseModel(IHost host, string name, CancellationToken cancellationToken = default)
+    {
+        if (await EnsureModelsInitialized(host, cancellationToken).ConfigureAwait(false))
```
```suggestion
        if (await EnsureModelsInitialized(host: host, performSelfcheck: false, cancellationToken: cancellationToken).ConfigureAwait(false))
```

cnupy reviewed (commented):
Thanks for the effort. Everything looks good to me! I just have a small suggestion to make the code logic easier to follow.

cnupy commented:
> the context field is deprecated and will be removed in a future version of Ollama

Yes, the correct way is to use the `chat` endpoint instead of the `generate` endpoint (ChatAsync method in the OllamaSharp client), as mentioned [previously](https://github.com/PowerShell/AIShell/pull/351#discussion_r2022295491).

daxian-dbw commented on shell/agents/AIShell.Ollama.Agent/Settings.cs:
```diff
@@ -1,40 +1,268 @@
+        }
+    }
+
+    private async Task<bool> EnsureModelsInitialized(IHost host, CancellationToken cancellationToken = default)
+    {
+        if (_initialized)
+        {
+            return true;
+        }
+
+        // Skip the self check when host is null.
+        if (host is not null)
+        {
+            bool success = await PerformSelfcheck(host, checkEndpointOnly: true);
+            if (!success)
+            {
+                return false;
+            }
+        }
```
`GetAllModels(IHost host = null, ...)` method has `null` as the default value for, and `host` is actually only used for endpoint check, which is not necessary in some cases, such as in `UseModel` action, `PerformSelfcheck` is called before calling into `EnsureModelsInitialized`.

How about having an overload that doesn't accept `host`? It would be clearer on the intention.

cnupy replied:
> `GetAllModels(IHost host = null, ...)` method has `null` as the default value for, and `host` is actually only used for endpoint check, which is not necessary in some cases, such as in `UseModel` action, `PerformSelfcheck` is called before calling into `EnsureModelsInitialized`.
> 
> ~How about having an overload that doesn't accept `host`? It would be clearer on the intention.~ Look at it more, I still think it's cleaner to skip endpoint check when `host is null`. This is because `EnsureModelsInitialized` is wrapped in utility methods such as `GetAllModels`, `UseModel`, `UsePreset` and etc, and we only know whether we need the endpoint check or not when calling those utility methods.
> 
> I added more comments to make the intent clear.

Sure, I'm OK with this.

daxian-dbw replied:
Great. I will get the PR merged.

daxian-dbw commented:
@cnupy Thank you for the contribution!
